proposals/0025-shader-execution-reordering.md

```diff
@@ -0,0 +1,1460 @@
+`ReorderThread` and shader table-based closesthit and miss shading to be combined with
+`RayQuery`.
+
+The proposed extension to HLSL should be relatively straightforward to adopt by
+current DXR implementations: `HitObject` merely decouples existing `TraceRay`
+functionality into two distinct stages: the traversal stage and the shading
+stage.
+For SER's `ReorderThread`, the minimal allowed implementation is simply a
+no-op, while implementations that already employ more sophisticated scheduling
+strategies are likely able to reuse existing mechanisms to implement support
+for SER. No DXR runtime changes are necessary, since the proposed extension to
+the programming model is limited to HLSL and DXIL.
+
+## Detailed Design
+
+This section describes the HLSL additions for `HitObject` and `ReorderThread`
+in detail.
+The canonical use of these features involve changing a `TraceRay` call to the
+following sequence:
+
+```C++
+HitObject Hit = HitObject::TraceRay( ..., Payload );
+ReorderThread( Hit );
+HitObject::Invoke( Hit, Payload );
+```
+
+This snippet traces a ray and stores the result of traversal, intersection
+testing and anyhit shading in `Hit`. The call to `ReorderThread` improves
+coherence based on the information inside the `Hit`. Closesthit or miss
+shading is then invoked in a more coherent context.
+
+Note that this is a very basic example. Among other things, it is possible to
+query information about the hit to influence `ReorderThread` with additional
+hints. See [Separation of ReorderThread and HitObject::Invoke](#separation-of-reorderthread-and-hitobjectinvoke)
+for more elaborate examples.
+
+### HitObject HLSL Additions
+
+```C++
+HitObject
+```
+
+The `HitObject` type encapsulates information about a hit or a miss. A
+`HitObject` is constructed using `HitObject::TraceRay`,
+`HitObject::FromRayQuery`, `HitObject::MakeMiss`, or `HitObject::MakeNop`. It
+can be used to invoke closesthit or miss shading using `HitObject::Invoke`,
+and to reorder threads for shading coherence with `ReorderThread`.
+
+The `HitObject` has value semantics, so modifying one `HitObject` will not
+impact any other `HitObject` in the shader. A shader can have any number of
+active `HitObject`s at a time. Each `HitObject` will take up some hardware
+resources to hold information related to the hit or miss, including ray,
+intersection attributes and information about the shader to invoke. The size
+of a `HitObject` is unspecified. As a consequence, `HitObject` cannot be
+stored in structured buffers. Ray payload structs must not have members of
+`HitObject` type. `HitObject` supports assignment (by-value copy) and can be
+passed as arguments to and returned from local functions.
+
+A `HitObject` is default-initialized to encode a NOP-HitObject (see `HitObject::MakeNop`).
+A NOP-HitObject can be used with `HitObject::Invoke` and `ReorderThread` but
+does not call shaders or provide additional information for reordering.
```

Comment:
> does not call shaders or provide additional information for reordering.

Should it be specified that both operations do nothing (`ReorderThread` causes no reordering)?  Or is it valid for an implementation to do thread reordering here when `HitObject` is a NOP-HitObject?

Reply:
I read it as that it is valid to do reordering, though no hint or any other info are available. So, likely `ReorderThread` does no reordering. But we don’t want to *guarantee* that it does nothing for the same reason that we don’t want to guarantee reconvergence with a `[NoReorder]` attribute.

Reply:
It is valid to reorder even with a NOP-HitObject. At the very least this makes sense if not all threads in a wave have a NOP-HitObject. Hints can still be used as well.

Reply:
@rasmusnv, I think this is a good explanation, thanks.